webhook/resources/snapshot/mutator.go

```diff
@@ -0,0 +1,93 @@
+package snapshot
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+	"github.com/pkg/errors"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	longhornFinalizerKey = longhorn.SchemeGroupVersion.Group
+)
+
+type snapShotMutator struct {
+	admission.DefaultMutator
+	ds *datastore.DataStore
+}
+
+func NewMutator(ds *datastore.DataStore) admission.Mutator {
+	return &snapShotMutator{ds: ds}
+}
+
+func (s *snapShotMutator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "snapshots",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.Snapshot{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+		},
+	}
+}
+
+func (s *snapShotMutator) Create(request *admission.Request, newObj runtime.Object) (admission.PatchOps, error) {
+	var patchOps admission.PatchOps
+
+	snapshot, ok := newObj.(*longhorn.Snapshot)
+	if !ok {
+		return nil, werror.NewInvalidError(fmt.Sprintf("%v is not a *longhorn.Snaphot", newObj), "")
+	}
+
+	volume, err := s.ds.GetVolumeRO(snapshot.Spec.Volume)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get volume %v", snapshot.Spec.Volume)
+		return nil, werror.NewInvalidError(err.Error(), "spec.Volume")
+	}
+
+	// Merge the user created and longhorn specific labels
+	labels := snapshot.Labels
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	for k, v := range types.GetVolumeLabels(volume.Name) {
+		labels[k] = v
+	}
+	bytes, err := json.Marshal(labels)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get JSON encoding for snapshot %v labels", snapshot.Name)
+		return nil, werror.NewInvalidError(err.Error(), "")
+	}
+	patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/metadata/labels", "value": %v}`, string(bytes)))
+
+	if err := util.AddFinalizer(longhornFinalizerKey, snapshot); err != nil {
+		return nil, werror.NewInvalidError(err.Error(), "")
+	}
+	bytes, err = json.Marshal(snapshot.Finalizers)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get JSON encoding for snapshot %v finalizers", snapshot.Name)
+		return nil, werror.NewInvalidError(err.Error(), "")
+	}
+	patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/metadata/finalizers", "value": %v}`, string(bytes)))
```

Comment:
Can replace the logic with the two new help functions https://github.com/longhorn/longhorn-manager/blob/master/webhook/common/common.go.

Reply:
Thank you. Fixed